modules/widgets/instagram/instagram.php

```diff
@@ -0,0 +1,397 @@
+			$result = $this->wpcom_json_api_request_as_blog( $path, 2, array( 'headers' => array( 'content-type' => 'application/json' ) ), null, 'wpcom' );
+			$response_code = wp_remote_retrieve_response_code( $result );
+			if ( 200 !== $response_code ) {
+				do_action( 'wpcomsh_log', 'Instagram widget: failed to verify nonce: API returned code ' . $response_code );
+				return 'ERROR';
+			}
+
+			$this->update_widget_token_id( $instance['token_id'] );
+		}
+		// If removing the widget's stored token ID
+		elseif ( $instance['token_id'] && isset( $_GET['instagram_widget_id'] ) && $_GET['instagram_widget_id'] == $this->number && ! empty( $_GET['instagram_widget'] ) && 'remove_token' == $_GET['instagram_widget'] ) {
+			if ( empty( $_GET['nonce'] ) || ! wp_verify_nonce( $_GET['nonce'], 'instagram-widget-remove-token-' . $this->number . '-' . $instance['token_id'] ) ) {
+				wp_die( __( 'Missing or invalid security nonce.', 'wpcom-instagram-widget' ) );
+			}
+			$site = Jetpack_Options::get_option( 'id' );
+			$path = sprintf( '/sites/%s/instagram/%s/remove', $site, $instance['token_id'] );
+			$result = $this->wpcom_json_api_request_as_blog( $path, 2, array( 'headers' => array( 'content-type' => 'application/json' ) ), null, 'wpcom' );
+
+			$response_code = wp_remote_retrieve_response_code( $result );
+
+			if ( 200 !== $response_code ) {
+				do_action( 'wpcomsh_log', 'Instagram widget: failed to remove keyring token: API returned code ' . $response_code );
+				return 'ERROR';
+			}
+
+			$instance['token_id'] = $this->defaults['token_id'];
+
+			$this->update_widget_token_id( $instance['token_id'] );
+		}
+		// If a token ID is stored, make sure it's still valid
+		elseif ( $instance['token_id'] ) {
+			$this->update_widget_token_id( $instance['token_id'] );
+		}
+
+		// No connection, or a legacy API token? Display a connection link.
+		if ( ! $instance['token_id'] ) {
+			echo '<p>' . __( '<strong>Important: You must first click Save to activate this widget <em>before</em> connecting your account.</strong> After saving the widget, click the button below to authorize your Instagram account.', 'wpcom-instagram-widget' ) . '</p>';
+			?>
+			<script type="text/javascript">
+				function getScreenCenterSpecs( width, height ) {
+					const screenTop = typeof window.screenTop !== 'undefined' ? window.screenTop : window.screenY,
+						screenLeft = typeof window.screenLeft !== 'undefined' ? window.screenLeft : window.screenX;
+
+					return [
+						'width=' + width,
+						'height=' + height,
+						'top=' + ( screenTop + window.innerHeight / 2 - height / 2 ),
+						'left=' + ( screenLeft + window.innerWidth / 2 - width / 2 ),
+					].join();
+				};
+				function openWindow() {
+					window.open(
+						'<?php echo $this->get_connect_url(); ?>',
+						'_blank',
+						'toolbar=0,location=0,menubar=0,' + getScreenCenterSpecs( 700, 700 )
+					);
+					window.onmessage = function( { data } ) {
+						console.log( data );
```

Comment:
The next step is to do an AJAX post to update the widget settings with this data...

Reply:
Took care of this in f432b7b!